spring-cloud-starter-stream-source-http/src/main/java/org/springframework/cloud/stream/app/http/source/HttpSourceConfiguration.java

```diff
@@ -41,28 +39,50 @@
  * @author Eric Bottard
  * @author Mark Fisher
  * @author Marius Bogoevici
+ * @author Artem Bilan
  */
-@Controller
 @EnableBinding(Source.class)
+@EnableConfigurationProperties(HttpSourceProperties.class)
 public class HttpSourceConfiguration {
 
 	@Autowired
 	private Source channels;
 
-	@RequestMapping(path = "${http.pathPattern:/}", method = POST, consumes = {"text/*", "application/json"})
-	@ResponseStatus(HttpStatus.ACCEPTED)
-	public void handleRequest(@RequestBody String body, @RequestHeader(HttpHeaders.CONTENT_TYPE) Object contentType) {
-		sendMessage(body, contentType);
+	@Autowired
+	private HttpSourceProperties properties;
+
+	@Bean
+	public HttpRequestHandlingEndpointSupport httpSourceString() {
+		return buildHttpRequestHandlerEndpointSpec("text/*", "application/json")
+				.requestPayloadType(String.class)
+				.get();
 	}
 
-	@RequestMapping(path = "${http.pathPattern:/}", method = POST, consumes = "*/*")
-	@ResponseStatus(HttpStatus.ACCEPTED)
-	public void handleRequest(@RequestBody byte[] body, @RequestHeader(HttpHeaders.CONTENT_TYPE) Object contentType) {
-		sendMessage(body, contentType);
+	@Bean
+	public HttpRequestHandlingEndpointSupport httpSourceBytes() {
+		return buildHttpRequestHandlerEndpointSpec("*/*")
+				.get();
 	}
 
-	private void sendMessage(Object body, Object contentType) {
-		channels.output().send(MessageBuilder.createMessage(body,
-				new MessageHeaders(Collections.singletonMap(MessageHeaders.CONTENT_TYPE, contentType))));
+	private HttpRequestHandlerEndpointSpec buildHttpRequestHandlerEndpointSpec(final String... consumes) {
+		// TODO Until SI Java DSL 1.2.2. Use mappedRequestHeaders() instead
+		DefaultHttpHeaderMapper defaultHttpHeaderMapper = DefaultHttpHeaderMapper.inboundMapper();
+		defaultHttpHeaderMapper.setInboundHeaderNames(this.properties.getMappedRequestHeaders());
+
+		return Http.inboundChannelAdapter(this.properties.getPathPattern())
+				.headerMapper(defaultHttpHeaderMapper)
+//				.mappedRequestHeaders(this.properties.getMappedRequestHeaders())
+				.statusCodeExpression(new ValueExpression<>(HttpStatus.ACCEPTED))
```

Comment:
I am wondering why do we needed that routing functionality in the first place, as publishing to the binder would reconvert the String to a byte[] anyway. Also, because the latter conversion is UTF-8 based and we send the header unchanged, doesn't it mean that we are republishing UTF-8 but potentially advertise a different encoding?